File: lexer.py
import re

# تعریف الگوها برای توکن‌ها
token_specs = [
    ('NUMBER',    r'\d+(\.\d+)?([eE][+-]?\d+)?'),  # عدد صحیح یا اعشاری یا علمی
    ('ID',        r'[a-zA-Z_][a-zA-Z0-9_]*'),      # شناسه
    ('EQ',        r'=='),                         # ==
    ('NE',        r'!='),                         # !=
    ('LE',        r'<='),                         # <=
    ('GE',        r'>='),                         # >=
    ('ASSIGN',    r'='),                          # =
    ('LT',        r'<'),                          # <
    ('GT',        r'>'),                          # >
    ('PLUS',      r'\+'),                         # +
    ('MINUS',     r'-'),                          # -
    ('MULT',      r'\*'),                         # *
    ('DIV',       r'/'),                          # /
    ('LPAREN',    r'\('),                         # (
    ('RPAREN',    r'\)'),                         # )
    ('LBRACE',    r'\{'),                         # {
    ('RBRACE',    r'\}'),                         # }
    ('SEMI',      r';'),                          # ;
    ('SKIP',      r'[ \t\n]+'),                   # فاصله و newline
    ('MISMATCH',  r'.'),                          # هرچیز نامعتبر
]

# لیست کلمات کلیدی
keywords = {'let', 'if', 'else', 'while', 'print', 'true', 'false', 'not', 'and', 'or'}

# کامپایل regex کلی
tok_regex = '|'.join(f'(?P<{name}>{pattern})' for name, pattern in token_specs)
get_token = re.compile(tok_regex).match

def tokenize(code):
    tokens = []
    pos = 0
    while pos < len(code):
        match = get_token(code, pos)
        if not match:
            raise SyntaxError(f'Unexpected character: {code[pos]}')
        kind = match.lastgroup
        value = match.group()

        if kind == 'SKIP':
            pass
        elif kind == 'ID' and value in keywords:
            tokens.append(('KEYWORD', value))
        elif kind == 'MISMATCH':
            raise SyntaxError(f'Illegal token: {value}')
        else:
            tokens.append((kind, value))
        pos = match.end()
    return tokens

if __name__ == '__main__':
    with open('input.txt', 'r') as f:
        code = f.read()
    tokens = tokenize(code)
    for token in tokens:
        print(token)

File: js_v/compiler/lexer.js
export function tokenize(input) {
  const keywords = new Set(['let', 'if', 'else', 'while', 'print', 'true', 'false', 'not', 'and', 'or']);
  const tokenSpecs = [
    ['NUMBER',    /\d+(\.\d+)?([eE][+-]?\d+)?/y],
    ['ID',        /[a-zA-Z_][a-zA-Z0-9_]*/y],
    ['EQ',        /==/y],
    ['NE',        /!=/y],
    ['LE',        /<=/y],
    ['GE',        />=/y],
    ['ASSIGN',    /=/y],
    ['LT',        /</y],
    ['GT',        />/y],
    ['PLUS',      /\+/y],
    ['MINUS',     /-/y],
    ['MULT',      /\*/y],
    ['DIV',       /\//y],
    ['LPAREN',    /\(/y],
    ['RPAREN',    /\)/y],
    ['LBRACE',    /\{/y],
    ['RBRACE',    /\}/y],
    ['SEMI',      /;/y],
    ['SKIP',      /[ \t\r\n]+/y],
    ['MISMATCH',  /./y]
  ];

  const tokens = [];
  let pos = 0;

  while (pos < input.length) {
    let match = false;
    for (const [type, regex] of tokenSpecs) {
      regex.lastIndex = pos;
      const result = regex.exec(input);
      if (result) {
        const value = result[0];
        match = true;

        if (type === 'SKIP') {
          pos += value.length;
          break;
        }

        if (type === 'ID' && keywords.has(value)) {
          tokens.push(['KEYWORD', value]);
        } else if (type === 'MISMATCH') {
          throw new SyntaxError(`Illegal token at position ${pos}: ${value}`);
        } else {
          tokens.push([type, value]);
        }

        pos += value.length;
        break;
      }
    }

    if (!match) {
      throw new SyntaxError(`Unexpected character at position ${pos}: '${input[pos]}'`);
    }
  }

  return tokens;
}
